### PR TITLE
ANDROID-651: Add json adapters for native enums

### PR DESCRIPTION
**Pull request by amytang0**

### CODE REVIEW

File: android/src/main/java/com/plaid/PlaidModule.kt [hunk 1]
```diff
@@ -15,12 +15,15 @@ import com.facebook.react.modules.core.DeviceEventManagerModule
...
 import com.google.gson.GsonBuilder
+import com.plaid.gson.*
 import com.plaid.link.Plaid
 import com.plaid.link.configuration.LinkLogLevel
 import com.plaid.link.configuration.LinkPublicKeyConfiguration
 import com.plaid.link.configuration.LinkTokenConfiguration
 import com.plaid.link.configuration.PlaidEnvironment
 import com.plaid.link.configuration.PlaidProduct
+import com.plaid.link.result.*
+import com.plaid.link.event.*
```
Comment (zsweigart) [inline android/src/main/java/com/plaid/PlaidModule.kt:26]: nit: * imports

File: android/src/main/java/com/plaid/gson/GsonTypeAdapters.kt [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+
+import com.google.gson.JsonDeserializationContext
+import com.google.gson.JsonDeserializer
+import com.google.gson.JsonElement
+import com.google.gson.JsonObject
+import com.google.gson.JsonPrimitive
+import com.google.gson.JsonSerializationContext
+import com.google.gson.JsonSerializer
+import com.plaid.link.result.*
+import com.plaid.link.event.*
```
Comment (zsweigart) [inline android/src/main/java/com/plaid/gson/GsonTypeAdapters.kt]: nit: * imports

File: android/src/main/java/com/plaid/gson/GsonTypeAdapters.kt [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+import java.lang.reflect.Type
+
+class AccountAdapter : JsonSerializer<LinkAccount> {
```
Comment (zsweigart) [inline android/src/main/java/com/plaid/gson/GsonTypeAdapters.kt]: nit: separate files

File: android/src/main/java/com/plaid/gson/GsonTypeAdapters.kt [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+
+  override fun serialize(
```
Comment (zsweigart) [inline android/src/main/java/com/plaid/gson/GsonTypeAdapters.kt]: since these are all in the Android SDK already, I'm worried copy-pasting them will cause issues if we forget to update here. 

They serialize/deserialize public classes anyway so could we move them to `com.plaid.link.internal.adapters` which would be next to `BasePlaid` so they aren't obfuscated but they are in and `internal` package and then commented to not use (although since they're gson adapters there's no real harm if people do use them)

**Review by zsweigart — CHANGES_REQUESTED** [review]
RC for copy-paste

TLDR; I think we can de-obfuscate the serializers in the core SDK and re-use them here

**Comment by zsweigart** [comment]
I think my comment got lost with the file split so https://github.com/plaid/react-native-plaid-link-sdk/pull/221#discussion_r505058847

File: android/src/main/java/com/plaid/gson/GsonTypeAdapters.kt (hunk 4 above)
Comment (amytang0) [reply android/src/main/java/com/plaid/gson/GsonTypeAdapters.kt]: Agreed, and we can tie the RN SDK to the correct Android SDK.

Fwiw, this is the one adapter that differs in RN vs Android (and making the Android ones visible will make that more apparent :))

Comment (JeroenMols) [reply android/src/main/java/com/plaid/gson/GsonTypeAdapters.kt]: We could express that in the adapter class name?

**Comment by amytang0** [comment]
Waiting on Android release of 3.0.2

**Review by zsweigart — APPROVED** [review]
